File: backend/app/Models/TradingPair.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;

class TradingPair extends Model
{
    use HasFactory;

    protected $fillable = [
        'symbol',
        'name',
        'type',
        'exchange',
        'package_access',
        'is_active',
        'pip_size',
    ];

    protected $casts = [
        'package_access' => 'array',
        'is_active'      => 'boolean',
        'pip_size'       => 'decimal:6',
    ];

    // Relationships
    public function signals(): HasMany
    {
        return $this->hasMany(Signal::class);
    }

    // Scopes
    public function scopeActive($query)
    {
        return $query->where('is_active', true);
    }

    public function scopeForPackage($query, string $packageSlug)
    {
        return $query->whereJsonContains('package_access', $packageSlug);
    }

    public function scopeCrypto($query)
    {
        return $query->where('type', 'crypto');
    }

    public function scopeForex($query)
    {
        return $query->where('type', 'forex');
    }

    // Helpers
    public function isAccessibleByPackage(string $packageSlug): bool
    {
        return in_array($packageSlug, $this->package_access ?? []);
    }

    public function isCrypto(): bool
    {
        return $this->type === 'crypto';
    }

    public function isForex(): bool
    {
        return $this->type === 'forex';
    }

    /**
     * Get the base and quote currencies.
     * e.g. EURUSD => ['EUR', 'USD'], BTCUSDT => ['BTC', 'USDT']
     */
    public function getCurrencyPair(): array
    {
        $symbol = $this->symbol;
        if ($this->isCrypto()) {
            if (str_ends_with($symbol, 'USDT')) {
                return [substr($symbol, 0, -4), 'USDT'];
            }
            if (str_ends_with($symbol, 'BTC')) {
                return [substr($symbol, 0, -3), 'BTC'];
            }
        }
        // Forex: 6-char symbols like EURUSD
        return [substr($symbol, 0, 3), substr($symbol, 3, 3)];
    }
}

File: android/app/src/main/java/com/chinarsignals/app/data/models/TradingPair.kt
package com.chinarsignals.app.data.models

import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize

@Parcelize
data class TradingPair(
    @SerializedName("id") val id: Int,
    @SerializedName("symbol") val symbol: String,
    @SerializedName("name") val name: String,
    @SerializedName("type") val type: String,
    @SerializedName("is_active") val isActive: Boolean,
    @SerializedName("base_currency") val baseCurrency: String? = null,
    @SerializedName("quote_currency") val quoteCurrency: String? = null
) : Parcelable {
    fun isForex(): Boolean = type.equals("forex", ignoreCase = true)
    fun isCrypto(): Boolean = type.equals("crypto", ignoreCase = true)
}
